### PR TITLE
stable hashing: Remove unused field and add documentation.

### PR DESCRIPTION
**Pull request by michaelwoerister**
This PR removes the `bytes_hashed` field from `StableHasher` which in the past has been used for collecting some statistics but has gone unused for quite a while (months at least) now.

The PR also tries to document some requirements for `HashStable` implementations that haven't been written down explicitly anywhere.

### CODE REVIEW

File: src/librustc_data_structures/stable_hasher.rs [hunk 1]
```diff
@@ -154,12 +135,34 @@ impl<W> Hasher for StableHasher<W> {
...
+/// - Stable hashes are sometimes used as identifiers. Therefore they must
+///   conform to the corresponding `PartialEq` implementations:
+///
+///     - `x == y` implies `hash_stable(x) == hash_stable(y)`, and
+///     - `x != y` implies `hash_stable(x) != hash_stable(y)`.
+///
+///   That second condition is usually not required for hash functions
+///   (e.g. `Hash`). In practice this means that `hash_stable` must feed any
+///   information into the hasher that a `PartialEq` comparision takes into
+///   account. See #49300 for an example where violating this invariant has
```
Comment (fbstj) [inline src/librustc_data_structures/stable_hasher.rs]:
```suggestion
///   account. See [#49300](https://github.com/rust-lang/rust/issues/49300) for an example where violating this invariant has
```

**Comment by estebank** [comment]
Other than the link for the sake of https://doc.rust-lang.org/nightly/nightly-rustc/rustc_data_structures/stable_hasher/trait.HashStable.html, r=me

File: src/librustc_data_structures/stable_hasher.rs (hunk 1 above)
Comment (michaelwoerister) [reply src/librustc_data_structures/stable_hasher.rs]: Done.

**Comment by michaelwoerister** [comment]
I updated the link.
@bors r=estebank rollup

**Pull request merged**